CIP-0119? | Governance Metadata - DReps (#788)
cardano-foundation/CIPs: CIP-0119/README.md

The [CIP-100 | Governance Metadata](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0100) standard provides a base framework for how all off-chain governance metadata should be formed and handled. But this is intentionally limited in scope, so that it can be expanded upon by more specific subsequent CIPs.

This proposal aims to provide a specification for off-chain metadata vocabulary that can be used to give context to CIP-100 for DRep registration and updates. Without a sufficiently detailed standard for DRep metadata we introduce the possibility to undermine the ability of DReps to explain their motivations and therefore people looking for someone to represent them with the best possible information available to make that choice. Furthermore a lack of such standards risks preventing interoperability between tools, to the detriment of user experiences.

#### Thank you
Thank you to [everyone who participated in the CIP workshops](#Acknowledgements)

## Motivation: why is this CIP necessary?

> CIP-1694 has set forth a compelling vision of a blockchain controlled by its community, and in doing so has thrown down the gauntlet to tooling providers to build apps and other kinds of tooling that will allow their users easy enjoyment of the governance features currently being built into the fabric of the Cardano blockchain. As I write this developers are feverishly hammering away at their keyboards like the blacksmiths of old to bring us the first iteration of these tools in a minimum viable format. 

Reviewer: ```suggestion
CIP-1694 has set forth a model of a blockchain controlled by its community, and in doing so has challenged tool providers to build apps and other kinds of tooling that will allow users easy access to the governance features currently being built into Cardano.  Minimum viable tools must be ready at the time these governance features are launched.
```
@Thomas-Upfield the paragraph as previously written violates a number of prohibitions we've long had in CIPs against:
- using first-person language
- writing from a personal point of view
- using subjective or dramatic language
- rooting CIPs in the time or the social conditions in which they are being written or expected to be merged / implemented.

Here is a redraft that I believe would address these problems, so please feel free to commit this version or equivalent <u>or</u> indicate what the dramatic and individualised language could accomplish behind simply identifying the requirements & time constraints of a minimum viable Governance launch.

Author: @rphair thanks for the rewrite. I have made the amendment.

This CIP has not included these features, the decision not to include these features was made in order to simplify this CIP so that it became suitable for the minimum viable level of tooling, with the expectation that further CIPs will build on it. 

### Witnesses
DRep Metadata will not follow the CIP-100 specification related to signing the metadata, the [authors property](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0100/README.md#high-level-description) can be left blank without the need for tooling providers to warn their users that the author has not been validated. Instead the author can be derived from the DRep ID associated with the registration. The need for an `authors` field will also be discarded in favour of including usernames inside of the `body` field. For the avoidance of doubt this CIP recommends that the entire authors property be left blank, and that tooling ignore it. 

### Extended Vocabulary

> Like CIP-108, this CIP also extends the potential vocabulary of CIP-100's `body` property. 

Reviewer 2: ```suggestion
Like CIP-108, this CIP also extends the potential vocabulary of CIP-100's `body` property. 

Furthermore we extend the Schema.org definition of a [Person](https://schema.org/Person). Any property of Person maybe included within the `body`.

  
```

Reviewer: @Ryun1 @Crypto2099 and perhaps highlight any of these potential dRep-relevant attributes of a [schema.org/Person](https://schema.org/Person):
- `memberOf`
- `worksFor`
- `jobTitle`
- `hasCredential`
- `hasOccupation`
- `affiliation`
- `colleague`

- expenses
- any incentive program

Therefore there MAY be a `paymentAddress` field in the metadata where such payments could be sent. This makes such an address public and payments to DReps transparent. 

#### `username`
- Compulsory
- A very short freeform text field. Limited to 80 characters.
- This MUST NOT support markdown text styling.
- It is intended that authors will use this field for their profile name/ username.

#### `picture`
- Optional 
- A base 64 encoded profile picture
- This SHOULD be treated as the profile picture of the individual

#### `objectives`
- Optional
- A freeform text field with a maximum of 1000 characters
- A short description of what the person believes and what they want to achieve as a DRep

#### `motivations`
- Optional
- A freeform text field with a maximum of 1000 characters
- A short description of why they want to be a DRep, what personal and professional experiences have they had that have driven them to register 

#### `qualifications`
- Optional
- A freeform text field with a maximum of 1000 characters 
- A space where the registrant can to list the qualifications they hold that are relevant to their role as a DRep

#### `references`

> - Compulsory

Reviewer 2: ```suggestion
- Compulsory
```
Why?

Reviewer: @Thomas-Upfield I am also curious "why" because of the reasons I mentioned at yesterday's CIP meeting:
- It might be well part of a human dRep's platform _not_ to have any of these references, or even to disavow such references.
- The dRep conception is currently aligned to human representatives but could be AIs, smart contracts aggregating other choices or behaviour, or other such abstracts soon enough... and these generally won't have references associated with human identity.

Author: I have changed this so that the only properties that are compulsory is the name property. This is the case because a person must have a name.

Reviewer: @Thomas-Upfield I would agree with the qualification that it's compulsory because an _entity_ must have a name.  I have still seen no compelling argument or proposed requirement (cc @Ryun1) that a dRep must be a person.

Furthermore we extend the Schema.org definition of a [Person](https://schema.org/Person). Any property of Person maybe included within the `body`.

>**Reminder for tooling providers/builders** DRep metadata is user generated content.

The following are a list of properties tooling should expect to encounter:

#### `paymentAddress`
- Optional

Dreps may want to recieve tokens for a variety of reasons such as:
- donations
- expenses
- any incentive program

Therefore there MAY be a `paymentAddress` field in the metadata where such payments could be sent. This makes such an address public and payments to DReps transparent. 

This SHOULD NOT be confused with the `address` property of a [Person](https://schema.org/Person), `address` in the context of a DRep refers to their location and NOT their payment address.

#### `givenName`
- Compulsory
- This is a property inherited from [Person](https://schema.org/Person)
- It is the only compulsory property
- A very short freeform text field. Limited to 80 characters.
- This MUST NOT support markdown text styling.
- It is intended that authors will use this field for their profile name/ username.

> #### `Image`

Reviewer 2: ```suggestion
#### `image`
```

- The creator of the metadata SHOULD add a `label`, this `label` SHOULD describe the source of the url, e.g. if it is a link to the DRep's X account then the `label` SHOULD say "X". If it is the only personal website provided by the DRep the `label` should say "Personal Website" rather than domain_name.com.
- The `label` of each `Link` SHOULD NOT be left blank
- Each `Link` MUST have exactly one `uri` (as specified in CIP-100) which SHOULD not be blank.

##### `type`: Identity
- Optional
- The `uri` of a reference with the `type` "Identity" is a way for DReps to prove that they are who they say they are
- It is expected that the "Identity" of a DRep will be the addresses of their most active social media account twitter, linkedin etc. or personal website.
- The DRep must reference their DRep ID in a prominent place at the location that is specified in the `uri` property of that reference. This will be used by people reviewing this DRep to prove and verify that the person described in the metadata is the same as the person who set up the social media profile.

#### `doNotList`

> - Optional

Reviewer 2: ```suggestion
- Optional
- Can be of value `true` or `false`.
```

### Witnesses
DRep Metadata will not follow the CIP-100 specification related to signing the metadata, the [`authors` property](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0100/README.md#high-level-description) can be left blank without the need for tooling providers to warn their users that the author has not been validated. Instead the author can be derived from the DRep ID associated with the registration. The need for an `authors` field will also be discarded in favour of including `givenName`s and `Identity` inside of the `body` field. For the avoidance of doubt this CIP recommends that the entire authors property be left blank, and that tooling ignore it. 

### Extended Vocabulary
Like CIP-108, this CIP also extends the potential vocabulary of CIP-100's `body` property. 

Furthermore we extend the Schema.org definition of a [Person](https://schema.org/Person). Any property of Person maybe included within the `body`.

>**Reminder for tooling providers/builders** DRep metadata is user generated content.

The following are a list of properties tooling should expect to encounter:

#### `paymentAddress`
- Optional

> - Bech32 encoded address

Reviewer 2: ```suggestion
- Bech32 encoded address, for the same network as the DRep registration is to be submitted to.
```

### Witnesses
DRep Metadata will not follow the CIP-100 specification related to signing the metadata, the [`authors` property](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0100/README.md#high-level-description) can be left blank without the need for tooling providers to warn their users that the author has not been validated. Instead the author can be derived from the DRep ID associated with the registration or update. The need for an `authors` field will also be discarded in favour of including `givenName`s and `Identity` inside of the `body` field. For the avoidance of doubt this CIP recommends that the entire authors property be left blank, and that tooling ignore it. 

### Extended Vocabulary
Like CIP-108, this CIP also extends the potential vocabulary of CIP-100's `body` property. 

Furthermore we extend the Schema.org definition of a [Person](https://schema.org/Person). Any property of Person maybe included within the `body`.

>**Reminder for tooling providers/builders** DRep metadata is user generated content.

The following are a list of properties tooling should expect to encounter:

#### `paymentAddress`
- Optional

> - Bech32 encoded address, for the same network as the DRep registration is to be submitted to.

Reviewer 2: ```suggestion
- Bech32 encoded payment address, for the same network as the DRep registration is to be submitted to.
```

> Dreps may want to recieve tokens for a variety of reasons such as:

Reviewer 2: ```suggestion
DReps may want to receive tokens for a variety of reasons such as:
```
spelling and capitalisation

#### `givenName`
- Compulsory
- This is a property inherited from [Person](https://schema.org/Person)
- It is the only compulsory property
- A very short freeform text field. Limited to 80 characters.
- This MUST NOT support markdown text styling.
- It is intended that authors will use this field for their profile name/ username.

#### `image`
- Optional
- This is a property inherited from [Person](https://schema.org/Person)
- This SHOULD be treated as the profile picture of the individual
- This MUST contain a fully described [`imageObject`](https://schema.org/ImageObject) property as opposed to a URL. 
 
##### `imageObject`
- This is to be included in a metadata file as a property of the `image` property, only if the `image` property is included.
- It explains the image to those (inc. tools) who are viewing it.
- `imageObject` MUST contain a base64 encoded image in its [`contentURL`](https://github.com/schemaorg/schemaorg/issues/2696) property in a [dataURI](https://en.wikipedia.org/wiki/Data_URI_scheme) format:
  - i.e. _data:content/type;base64,_ (AND NOT _data:domain.tld_)

>   - e.g. _contentURL:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==_ (AND NOT _contentURL:https://avatars.githubusercontent.com/u/113025685?v=4_)

Reviewer 3: I dislike this requirement that the image must be embedded within the metadata file itself as this will quickly lead to relatively massive (multi-megabyte) metadata files that are more difficult to fetch and store without gaining substantial value.

The schema.org `imageObject` already provides for a `sha256` field to provide a confirmation and validation hash that the asset in question is unmodified from the time of submission and allows for offloading. This will become particularly important when (most likely in the future) we have groups of dReps who all have a profile photo or we have proposals that need quite a few different images as citation and reference. 

It is more logical to develop a strategy to work with "remote" files rather than require that they all be b64 encoded and included directly in the metadata file itself.

Reviewer 3: I would modify the language such:

**If the `imageObject` DOES NOT contain a base64 encoded image, the `sha256` property MUST be populated and match the contents of the image file found at the `contentUrl`**

Author: I get where you are coming from, have you considered:

- The point of having external metadata files (e.g. on IPFS) was that people could store more information without congesting Cardano
- Introduction of another hashing algorithm other than blake 2b could be confusing
- Another method of saving images would add complexity for tooling providers who are building tools that want to show these images
  - If tools reading the metadata need to visit the url link every time they show an image this would take ages for the image to load. 
  - Alternatively they could save images and only show them if the SHA256 hash matches 
- If tools want to implement a porn filter b64 images only need to be checked once whereas an image at a url might change and so would need to be checked every time they are opened. This is somewhat ameliorated by the hash check, but again this is more overhead for tooling.

?

Author: On the other hand CIP-100 does allow for metadata to be saved as part of the governance transaction so maybe your suggestion is necessary for such instances. I'm keen for a discussion.

Author: @Crypto2099 , @rphair , @Ryun1  I have made changes in line with the above comments.
